src/benchmark/tpch_benchmark.cpp

```diff
@@ -0,0 +1,355 @@
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <string>
+
+#include "cxxopts.hpp"
+#include "json.hpp"
+#include "sql/sql_pipeline.hpp"
+#include "tpch/tpch_db_generator.hpp"
+#include "tpch/tpch_queries.hpp"
+#include "SQLParser.h"
+#include "SQLParserResult.h"
+
+/**
+ * This benchmark measures Hyrise's performance executing the TPC-H queries, it doesn't (yet) support running the TPC-H
+ * benchmark as it is specified.
+ * The benchmark offers a wide range of options (scale_factor, chunk_size, ...) but most notably it offers two modes:
+ * IndividualQueries and PermutedQuerySets. See docs on BenchmarkMode for details.
+ * The benchmark will stop issuing new queries if either enough iterations have taken place or enough time has passed.
+ *
+ * main() is mostly concerned with parsing the CLI options while TpchBenchmark::run() performs the actual benchmark
+ * logic.
+ */
+
+namespace opossum {
+
+/**
+ * IndividualQueries runs each query a number of times and then the next one
+ * PermutedQuerySets runs the queries as sets permuting their order after each run (this exercises caches)
+ */
+enum class BenchmarkMode { IndividualQueries, PermutedQuerySets };
+
+using Duration = std::chrono::high_resolution_clock::duration;
+using TimePoint = std::chrono::high_resolution_clock::time_point;
+
+// Used to config out()
+auto verbose_benchmark = false;
+
+/**
+ * @return std::cout if `verbose_benchmark` is true, otherwise returns a discarding stream
+ */
+std::ostream& out() {
+  if (verbose_benchmark) {
+    return std::cout;
+  }
+
+  // Create no-op stream that just swallows everything streamed into it
+  // See https://stackoverflow.com/a/11826666
+  class NullBuffer : public std::streambuf {
+   public:
+    int overflow(int c) { return c; }
+  };
+
+  static NullBuffer null_buffer;
+  static std::ostream null_stream(&null_buffer);
+
+  return null_stream;
+}
+
+struct QueryBenchmarkResult {
+  size_t num_iterations = 0;
+  Duration duration = Duration{};
+};
+
+using BenchmarkResults = std::unordered_map<size_t, QueryBenchmarkResult>;
+
+/**
+ * Loosely copying the functionality of benchmark::State
+ * keep_running() returns false once enough iterations or time has passed.
+ *
+ */
+struct BenchmarkState {
+  enum class State { NotStarted, Running, Over };
+
+  BenchmarkState(const size_t max_num_iterations, const Duration max_duration)
+      : max_num_iterations(max_num_iterations), max_duration(max_duration) {}
+
+  bool keep_running() {
+    switch (state) {
+      case State::NotStarted:
+        begin = std::chrono::high_resolution_clock::now();
```

Comment:
It seems weird to me that a method that returns a bool "keep running" should be in charge of taking the time.

Reply:
That’s actually taken from Google Benchmark ;)

Reply:
And those are the guys that did the `Range` thing, so that doesn't count for much. ;)

Reply:
I think the `KeepRunning()` "pattern" is not beautiful, but it works. Allows for one construct (a while loop) to control and count the number of iterations as well as measuring exact begin and end times.

Reply:
works for me